cmake: add and use new DT alias helpers

mbolivar-nordic opened the pull request:
Add convenience functions for dealing with devicetree aliases to extensions.cmake. Document them properly and mention them in the release notes. Document that they are expected to be added on an as-needed basis.

mbolivar-nordic committed:
cmake: extensions: add target_sources_if_dt_node()

This is like target_sources_ifdef(), except it works if a devicetree
node exists.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -1472,6 +1472,13 @@ function(target_sources_ifdef feature_toggle target scope item)
   endif()
 endfunction()
 
+function(target_sources_if_dt_alias alias target scope item)
```
Could you please add some description of this function.

I'm curios on one thing.
This adds a `_if_dt_alias` which internally uses `dt_prop(... "/aliases" ...)` but where will this end then ?
Will there be a dedicated function for each path / property someone might ask for ?

Like:
`target_sources_if_dt_status_ok()` calling `dt_prop(... PATH <path> STATUS ...)

or other dt_prop dependent calls with other combinations

I think I understand the benefit, but what i'm trying to understand is how many places you expect this to be used, or if  this function should be a bit more generic so we won't need to add `n` dt functions for each combination of `dt_prop()`.

Maybe:
`target_sources_if_dt_prop(<path> <prop> <target> <scope> <item>)`
which in your case could be invoked as:
`target_sources_if_dt_prop("/aliases" <prop> <target> <scope> <item>)` (`<prop>` being your current `<alias>`)

mbolivar-nordic replied:
I have reworked this PR substantially to address your concerns. Please revisit.

mbolivar-nordic committed:
cmake: extensions: DT helper fixes

Fix miscellaneous issues:

- dt_node_has_status() is missing a return() statement
- dt_chosen()'s comment string and error handling message contain errors

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
cmake: extensions: add dt_alias()

This is a helper function for looking up the path to a devicetree
alias. It is analogous to dt_chosen(). It has to exist as a separate
function from dt_prop() for similar reasons that dt_chosen() does:
the edtlib APIs for interacting with /aliases are special-cased in the
same way they are for /chosen.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_dts_cmake: a small optimization

Using edt.chosen_nodes looks like a simple attribute lookup, but it's
actually calling a property that will create a new list of chosen
nodes every time. Apply a small optimization by only creating the list
once.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
scripts: gen_dts_cmake: emit properties for /aliases

This enables cmake extensions that can look up the path for any
devicetree alias, or check if the alias is missing, etc.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
doc: release-notes: mention new cmake DT extensions

Mention some new devicetree helpers and changes to existing ones.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
I was firing this off quickly yesterday to get initial reviews. Thanks for those.

Since @tejlmand had questions about scope and ambition, I've spent more time trying to produce a more thorough PR. Hopefully it is acceptable now. Please revisit your reviews.

pdgendt commented on cmake/extensions.cmake:
```diff
@@ -2940,6 +2943,104 @@ function(dt_chosen var)
+#
+# <alias>     : Devicetree alias to check
+# ...         : Additional arguments are passed to message()
+function(message_if_no_dt_alias alias)
```
This function is now used in the samples, which previously verified the `status = okay` property. Does `dt_alias` also take that property into account?

tejlmand reviewed (changes requested):
I think `dt_alias()` should support `REQUIRED` flag instead, like:
```
dt_alias(<var> ALIAS <alias> REQUIRED)
```

Also, from the updates in this PR, then there is not really a need for `target_sources_if_dt_alias()` as all the places it is being used are not really having optional inclusion of sources based on devicetree alias, all places require the devicetree alias being present.

mbolivar-nordic committed:
cmake: extensions: add internal DT helpers

Add a couple of helper functions we'll use to make some other
improvements.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic committed:
cmake: extensions: improve DT consistency with python APIs

In the python APIs we use for devicetree, you can generally use an
alias interchangeably with the full path to a node.

For example, dtlib.py says:

    [...] both dt.get_node("/foo/bar") and dt.get_node("bar-alias")
    will return the 'bar' node below:

      /dts-v1/;

      / {
              foo {
                      bar_label: bar {
                              baz {
                              };
                      };
              };

              aliases {
                      bar-alias = &bar-label;
              };
      };

    Fetching subnodes via aliases is supported:
    dt.get_node("bar-alias/baz") returns the 'baz' node.

Now that we have the necessary helper functions for resolving alias
names, make the CMake APIs to devicetree behave the same way. This
improves consistency between the two sets of devicetree helpers used
by the build system, making them both match the general devicetree
practice that an alias is as good as a path. This is good for
consistency.

Signed-off-by: Martí Bolívar <marti.bolivar@nordicsemi.no>

mbolivar-nordic commented:
The latest rework follows some more discussion with @tejlmand offline.

- all changes to samples removed; this can't work with twister as-is and it's too much work to make this situation better
- we now treat aliases interchangeably with full paths, for consistency with the general DT practice as well as the python APIs
- message_if_dt_alias removed
- target_sources_if_dt_alias is now target_sources_if_dt_node, which takes a path that may be an alias

tejlmand commented on cmake/extensions.cmake:
```diff
@@ -2990,6 +2990,77 @@ function(dt_chosen var)
+#   dt_path_internal(ret "my-alias/baz") # sets ret to "/foo/bar/baz"
+#   dt_path_internal(ret "/blub")        # ret is undefined
+function(dt_path_internal var path)
+  string(FIND "${path}" "/" slash_index)
```
an alternative would be to use a `REGEX MATCH` like this:
```
string(REGEX MATCH "^([/])?([^/]+)(/.*)*" match_result "${PATH}")
```

For example that will give:
```
path: /foo                                                                                                                                                                                                                                   
CMAKE_MATCH_1: /                                                                                                                                                                                                                             
CMAKE_MATCH_2: foo                                                                                                                                                                                                                           
CMAKE_MATCH_3:                                                                                                                                                                                                                               

path: /foo/bar                                                                                                                                                                                                                               
CMAKE_MATCH_1: /                                                                                                                                                                                                                             
CMAKE_MATCH_2: foo                                                                                                                                                                                                                           
CMAKE_MATCH_3: /bar                                                                                                                                                                                                                          

path: my-alias                                                                                                                                                                                                                               
CMAKE_MATCH_1:                                                                                                                                                                                                                               
CMAKE_MATCH_2: my-alias                                                                                                                                                                                                                      
CMAKE_MATCH_3:                                                                                                                                                                                                                               

path: my-alias/baz                                                                                                                                                                                                                           
CMAKE_MATCH_1:                                                                                                                                                                                                                               
CMAKE_MATCH_2: my-alias                                                                                                                                                                                                                      
CMAKE_MATCH_3: /baz      
```

So you can use `CMAKE_MATCH_1` to identify if `path` starts with `/`.
`CMAKE_MATCH_2` as `alias_name` and if `CMAKE_MATCH_3` is defined, then it's an alias path.

Doing so could cleanup some logic below I believe.

mbolivar-nordic replied:
> an alternative would be to use a REGEX MATCH like this:

That regular expression doesn't work for the root node, `/`. I would prefer to leave this as-is.

tejlmand replied:
> That regular expression doesn't work for the root node, `/`.

I wasn't aware the root node was a use-case also.
In that case the regex should look as
```
string(REGEX MATCH "^([/])?([^/]*)(/.*)*" match_result "/") 
```

> I would prefer to leave this as-is.

I'm fine with that as well.

mbolivar-nordic replied:
>> I would prefer to leave this as-is.
>
> I'm fine with that as well.

OK, that was your last unresolved comment, then, I think. OK to approve?

tejlmand commented on doc/releases/release-notes-3.0.rst:
```diff
@@ -287,6 +287,21 @@ USB
 Build and Infrastructure
 ************************
 
+* Build system
```
Nothing wrong on this line, but commit message still mentions: `message_if_no_dt_alias()`.

> Mention the dt_alias(), target_sources_if_dt_alias(), and
message_if_no_dt_alias() helpers.

tejlmand reviewed (commented):
Generally good looking :+1: 

Some minor nits, and an alternative regex proposal.

mbolivar-nordic commented on cmake/extensions.cmake:
```diff
@@ -2990,6 +2990,77 @@ function(dt_chosen var)
+    if (DEFINED alias_path)
+      set(rest)
+      if (NOT "${slash_index}" EQUAL -1)
+	string(SUBSTRING "${path}" "${slash_index}" -1 rest)
```
Whoops, looks like I am using tabs here, will fix.

mbolivar-nordic commented:
@tejlmand thanks for review, comments addressed

tejlmand commented:
> @tejlmand thanks for review, comments addressed

except your last push didn't fix the indentation.
https://github.com/zephyrproject-rtos/zephyr/compare/db6a9c39e83487594d8a307e2eabc5e720150634..28f3cb9d85d62b668b6b82b9230b791b6483b59f

mbolivar-nordic commented:
> except your last push didn't fix the indentation.

Darn, sorry, I threw away the fix when I realized the regex wouldn't handle the root node. I got it this time, thanks for catching that.
